### ISSUE TITLE
Add Block: Title

### ISSUE THREAD

**Issue by DevOnWheels** [WordPress/gutenberg#11553]
Hey, this is just a question/ idea: 
I tried to insert HTML in my posts title. It does work, but the HTML is actually written out in the edit field of the title. This looks awkward. Is there any possibility to interpret the HTML in post titles in the same way as in the rest of Gutenberg. Like some kind of a "post title block"?

Reason for the question is that I would like to have overlines in my h1 tag without using ACF for that purpose. If this tracker is not the place for ideas/ enhancements, just drop me a note and close the ticket :)

Best regards, Matt

**Comment by designsimply** [comment]
Great question! There's some related discussion in #8469 where it is mentioned in the comments that future plans include making the title area into a block which likely means you would be able to add HTML to it. I checked the `New Block` label and didn't see one made for that specifically though, so I'm updating the title here so we can use this request for tracking. Thanks for raising the issue!

**Comment by designsimply** [comment]
Related: #8824 (and noting a kicker block may not be needed if title becomes a block that can be reordered within a post)

**Comment by designsimply** [comment]
Noting another relevant quote about title (and other things) eventually becoming blocks:

> Eventually, the title, the date, the author, etc, will all just be blocks you can place in whatever order around the content.

Source: https://github.com/WordPress/gutenberg/issues/4482#issuecomment-357705649

**Comment by mapk** [comment]
I'd really like to see the title block come to fruition in Phase 2. I'll add it to the milestone. We'll need some design input on this though... for instance, where does the meta info that the title provided go?

**Comment by DevOnWheels** [comment]
> I'd really like to see the title block come to fruition in Phase 2. I'll add it to the milestone. We'll need some design input on this though... for instance, where does the meta info that the title provided go?

@mapk what exactly do you mean by "meta info that the title provided"? Happy to help :)

**Comment by mapk** [comment]
> what exactly do you mean by "meta info that the title provided"? Happy to help :)

The permalink is what I mean here. It comes from the Title.

![screen shot 2019-01-08 at 2 38 02 pm](https://user-images.githubusercontent.com/617986/50863302-18280500-1353-11e9-830c-ccb396f6d49e.png)

Would the Title block still incorporate the permalink info as it does now? Would that move to the sidebar? Just some things to consider.

**Comment by DevOnWheels** [comment]
Yes, the current place is the "correct" one, I would say. A few thoughts on the "title block" here:
The block should be available only once, so there cannot be confusion between different titles. Additionally I would like to have a field for a "working title" so that I can have a short title in my admin lists, but a long title with HTML etc. in the title block. For example "This is my <strong>super awesome</strong> frontpage! Check it out!" in the block, but only "Static Frontpage" in the page list. And for these reasons I would like to have the settings directly at the block.

**Comment by mapk** [comment]
> The block should be available only once, so there cannot be confusion between different titles. 

Definitely possible. But what if the user doesn't want a title block on their page? Then how do we define the URI?

**Comment by DevOnWheels** [comment]
Good question. This would be a reason to exclude the settings to the options bar of the document. On the other hand, I really get headache about the fact that an article wouldn't have a title, but that's a SEO problem and not related to this issue.

**Comment by swissspidy** [comment]
> Definitely possible. But what if the user doesn't want a title block on their page? Then how do we define the URI?

WordPress uses the post ID for the URL by default if there's no title set. Unless the user manually overrides the permalink with their own liking.

**Comment by DevOnWheels** [comment]
> WordPress uses the post ID for the URL by default if there's no title set. Unless the user manually overrides the permalink with their own liking.

Exactly, but if you don't have the title block in your document (and the permalink option would be tied to the block settings) you will not have the settings either.

**Comment by mapk** [comment]
A common problem I've experienced is the homepage. I want a homepage, and want the URL to reflect `/home` or `/welcome` in the URL, but I don't want to have a title that says that... and I definitely don't want a random number.

**Comment by DevOnWheels** [comment]
@mapk a solution then would be to put the settings of the title into the document tab of the document so that the user doesn't have to define a title for the document at all.

**Comment by mapk** [comment]
> put the settings of the title into the document tab

I think this is the best solution right now and feel the document tab is a perfect fit for it, but I'd love to hear if others have any more suggestions.  Also, @DevOnWheels, are you able to mock this up for us?

**Comment by jasmussen** [comment]
One idea to consider. If the block can be deleted, then we need a different way to change the title of a post. Google Docs, Dropbox Paper, and many others, show the title in the editor bar, and auto-fill it based on the contents of the post. But with a deletable title block, this would still allow you to customize it. This might also afford us an interface for more intuitively customize the permalink slug. 

CC: @paaljoachim, https://github.com/WordPress/gutenberg/issues/4573#issuecomment-455805418

**Comment by paaljoachim** [comment]
It seems the discussion is going something like this:

The existing title box:
![screen shot 2019-01-21 at 10 10 06](https://user-images.githubusercontent.com/5323259/51464469-ead15300-1d65-11e9-94d4-eebb7284b3f5.png)

Removing the title:
![screen shot 2019-01-21 at 10 10 40](https://user-images.githubusercontent.com/5323259/51464495-fb81c900-1d65-11e9-9b2c-f751ac9433f2.png)

Now lets say that we removed the title block.
The title defaults to the top bar.

![gutenberg-title-top-bar](https://user-images.githubusercontent.com/5323259/51464553-2704b380-1d66-11e9-8dfa-edb39b730113.jpg)


![gutenberg-title-top-bar-2-adding-title](https://user-images.githubusercontent.com/5323259/51464582-32f07580-1d66-11e9-9874-d9fa7250bd16.jpg)


![gutenberg-title-top-bar-3-added-title-permalink-bar](https://user-images.githubusercontent.com/5323259/51465151-90d18d00-1d67-11e9-8aa2-3cddc9d98fd7.jpg)

In the permalink I removed the square with a arrow in it. Click it to view the page. (I just find that icon confusing as there are more consistent ways to preview the page. Like clicking preview or view page .) I also removed the edit button. As one should be able to click the permalink to change it as was done in earlier classic editor.

**Comment by paaljoachim** [comment]
Well on smaller screens the permalink could move below the title. Also remove the word Permalink.
Title is selected and the permalink link shows up below it. One should be able to click the edit button on larger screens OR the link directly to select it and adjust. On smaller screens the edit button can be removed as the permalink area adds a border around it signalling that one can click into it as any other field.

![gutenberg-title-bar-responsive](https://user-images.githubusercontent.com/5323259/51466014-84e6ca80-1d69-11e9-89d7-6e363d1f789d.jpg)

**Comment by mapk** [comment]
Thanks to @paaljoachim's mockups, I just noticed an inspector accordion that I never realized was there: **Permalink**.

Is it possible to just remove the link entirely from the Title Block in favor of the inspector interface? We can even add the page title there as well. 

![permalink](https://user-images.githubusercontent.com/617986/51642031-c8356a80-1f1c-11e9-97c5-ab1ca3a2063b.png)

**Comment by paaljoachim** [comment]
The question that comes up for me is what is the main purpose of removing the title block?

**Comment by mapk** [comment]
> The question that comes up for me is what is the main purpose of removing the title block?

I think a good example is on homepages. I don't want to a big title at the top that says, "HOME" or "WELCOME".

> Does this mean it could be visible in the backend but not in the frontend?

It could, yes. But it seems to become just meta data at that point. In which case, wouldn't be better suited in the inspector?

> The permalink is an important feature and should be clearly visible somehow in the layout area or above it.

I agree, but it's an important feature for the document, not for a particular block. It's not really _needed_ for the Title blocks functionality as content.

I do like the ideas of adding it to the top bar similarly to your mockups, @paaljoachim. My question is, is it important enough to validate its own spot in the top bar? Maybe it is?

**Comment by dannycolin** [comment]
> I do like the ideas of adding it to the top bar similarly to your mockups [...] People probably expect to see it there.

If the title is hidden in the sidebar with the permalink a lot of users will probably be confused. 

The only problem I can see with the title in the top bar is on mobile especially in portrait mode. It will be too much truncated to be useful. For that use case, adding a *title* field in the Permalink (maybe renaming it *Title & Permalink*) would makes more sense.

**Comment by jasmussen** [comment]
Expanding on https://github.com/WordPress/gutenberg/issues/11553#issuecomment-455992928 and @paaljoachim's idea of showing the title in the editor bar, here's a sketch inspired by Google Docs and Dropbox Paper. 

![screenshot 2019-01-24 at 08 40 20](https://user-images.githubusercontent.com/1204802/51662335-ca96d380-1fb3-11e9-99aa-a5fdab0b7ac7.png)

It remains to be seen how feasible that particular implementation is — it asks for a bit more space in that topbar, which won't be present when the top toolbar is enabled. All of that is fixable, but it means we have to think more about it.

But the problems it solves are these:

- It allows you to edit the page/post title even if you've deleted the title block
- It could allow you to have a different page title than the first heading on the page (mindblown.webm)
- It could hold an interface that _feels_ (worth prototyping maybe) like it would be an obvious place to also hold a permalink editing interface

@dannycolin's point is pertinent — this does not scale well to mobile in the current configuration. But that's fixable too — we could do what Google Docs has, which is a preview mode with one set of buttons in the toolbar, including the title, and an edit mode, with another set of buttons in the toolbar. Or we can leverage sticky positioning to have two stacked toolbars, one which scrolls out of view when you thumb down the page. Or we could just _not show it_ on mobile, and show the title on the post-publish experience which is enabled on all mobile breakpoints.

**Comment by paaljoachim** [comment]
Here are examples of wide screen and narrow screens with Google Doc and Microsoft Word Online.

A wide Google Doc screen.
![screen shot 2019-01-24 at 10 18 56](https://user-images.githubusercontent.com/5323259/51668098-c07bd180-1fc1-11e9-9263-7d4b560e7104.png)

A narrow Google Doc screen.
![screen shot 2019-01-24 at 10 16 43](https://user-images.githubusercontent.com/5323259/51668154-dc7f7300-1fc1-11e9-9534-2ee2970430fd.png)


A wide Word screen
![screen shot 2019-01-24 at 10 18 44](https://user-images.githubusercontent.com/5323259/51668172-e6a17180-1fc1-11e9-865b-58d2bd76113b.png)

A narrow Word screen
![screen shot 2019-01-24 at 10 19 35](https://user-images.githubusercontent.com/5323259/51668190-edc87f80-1fc1-11e9-9a3f-32aab31de633.png)


Both prioritize which options are visible at smaller resolutions.

**Comment by paaljoachim** [comment]
I would think that the principle would be the same even though it is meant for the mobile app.
As in what options need to be prioritized with the mobile app compared to the standard desktop screen. As well as how to show options at very small screen sizes. As in click into the title to see the permalink above or below. It will then use the full 320px wide area. If that is too small for the permalink what about going from showing https://sitename.com/parentpage/titleofpage to just
titleofpage. 

Here is the responsive view at minimum size at 411px on my desktop:

![gutenberg-narrow-responsive-view-permalink](https://user-images.githubusercontent.com/5323259/51670040-0dfa3d80-1fc6-11e9-8b9c-3704fbcb1095.jpg)

**Comment by ChemicalSailor** [comment]
Just wanted to show my support for these explorations. I raised something similar a while ago in [#1914](https://github.com/WordPress/gutenberg/issues/1914) but it got deferred out of phase 1, glad to see it's being looked at again. It would be worth looking back at that thread, there were a few ideas.

**Comment by justintadlock** [comment]
Just sharing a use case for this below.  What I'm currently doing is creating a "No Title" page template and having the user add a H1 Heading block to handle this.  It'd particularly useful to be able to move the post title into something like a Cover block.

![ex-074](https://user-images.githubusercontent.com/1816309/56660134-78be2200-6664-11e9-9f47-046d033f708c.png)

---

One thing we must be mindful of though is that some themes will also have a byline or meta area for their posts before/after the title.  When building a post title block, those common elements must also be considered.  If the title can be moved around, what happens to those areas?

**Comment by afercia** [comment]
Re: the permalink UI, see previous #5494 for some important accessibility concerns.

Worth reminding the current permalink UI [should be considered deprecated](https://github.com/WordPress/gutenberg/issues/5494#issuecomment-439104668), see [also this other comment](https://github.com/WordPress/gutenberg/issues/10996#issuecomment-439097065). It's candidate for removal since a few months, in favor of the permalink UI in the sidebar.

Alternatively, if the final decision will be different and the current UI will stay, all the accessibility problems from #5494 and #15298 need to be solved. Especially #5494 outlines some important problems related to the placement of this UI in the DOM and the tab order, originally reported on March 2018 and waiting to be solved since more than one year

**Comment by paaljoachim** [comment]
Calypso title issue:
Full Site Editing: Site Title Block [8]
https://github.com/Automattic/wp-calypso/issues/33106

**Comment by sarahmonster** [comment]
Let's step back and take a look at our goals & some possible solutions. 

## What do we need of a title block?
- Ability to show & hide the block in post content.
- Need to be able to add text & blocks both above & below.
- Limit to a single title block per page.
- Permalink is created based on the title input, but doesn't necessarily need to be connected to the title block.
- Users may want to have a different display title than the actual title of the page.

I'm going to explore this starting from the mobile experience, in the interests of keeping things constrained and focused on logical source order. 

As far as I see this breaking down, we have three potential areas where we could reveal the post title:

## Option A

The title block is just a slightly-customised Heading block:

<img width="676" alt="Gutenberg - 2019-07-11 23 54 57" src="https://user-images.githubusercontent.com/376315/61142939-de35e800-a4c8-11e9-83d6-e120fb973d3b.png">

In it, we provide the option to hide or display the block. A "hidden" title block shows up in the post content (so it can be selected and manipulated), but in greyed-out manner to indicate it's not visible in the front-end. 

This affords a lengthy title nicely and plays well on mobile, as well as playing to Gutenberg's principle of putting as much as possible into the editor interface itself. However, I'm not certain that subverting the standard model of deleting blocks makes sense here. Let's explore what it would look like if instead of greying out, a deleted title block disappeared entirely. How would a user see and edit their document title?

## Option B

One option, as explored above, it to show the title in the page toolbar:

<img width="977" alt="Gutenberg - 2019-07-12 16 16 15" src="https://user-images.githubusercontent.com/376315/61142948-e4c45f80-a4c8-11e9-82c6-93bcbc353c3c.png">

This is definitely tricky to do on a mobile device—the best solution seems to be to create a new line for the title to stand on its own, which could potentially make for a noisy interface, especially when the title is immediately visible below for the majority of users. 

For consistency, we'd probably want to show the title here whether there was a title block present on the page or not, but it does create some duplication. And should the title be user-editable from here? This feels like a good place to _show_ the post title, but maybe not edit it, especially since it's divorced from any other edit controls, which exist either in the post content itself, or in the Settings Sidebar.

## Option C

Use the Document Settings to allow for a document title entry:

<img width="582" alt="Gutenberg - 2019-07-12 17 14 40" src="https://user-images.githubusercontent.com/376315/61143507-5224c000-a4ca-11e9-9779-6574e6af29b2.png">

This feels like an extremely natural solution, since it groups the document title with other document-level settings. Pairing it with the permalink helps to reinforce that connection for users who are accustomed to the title + permalink pairing, without isolating new users who haven't learnt this pattern already. 

It also may serve to make permalink settings a bit more findable than they currently are, but the Settings Sidebar is less visible on mobile overall.

## Mystery option D: a combination of the above

I suspect our best solution to the problem at hand will be a combination of the above—revealing the document title somewhere in the UI (probably the sidebar) and loosely associating it with an optional "display title" block. Revealing the permalink more prominently in the Document Settings feels like a step in the right direction here, allowing us to disconnect the permalink (document setting) from the display title (post content), and helping resolve some accessibility issues whilst opening up the title block for more block-like behaviour.

**Comment by dannycolin** [comment]
> @sarahmonster In it, we provide the option to hide or display the block. A "hidden" title block shows up in the post content (so it can be selected and manipulated), but in greyed-out manner to indicate it's not visible in the front-end.

Just a quick question related to the headings hierarchy and web accessibility. When you said "hidden" does it mean the Title block is removed from the DOM and the page/post will have no <h1> declared (unless the theme has one in the template file)?

**Comment by afercia** [comment]
> I'm going to explore this starting from the mobile experience, in the interests of keeping things constrained and focused on logical source order.

Thanks for this, greatly appreciated as it helps a lot accessibility as well ❤️Could be material for an entire talk 🙂

**Comment by sarahmonster** [comment]
> Just a quick question related to the headings hierarchy and web accessibility. When you said "hidden" does it mean the Title block is removed from the DOM and the page/post will have no <h1> declared?

I'm a bit unsure on this detail right now—my instinct is that the h1 should definitely still appear in the DOM, since users probably wouldn't replace it with a suitable h1 for the page, but that could be problematic if they _do_ use an h1 elsewhere as a title. Maybe this behaviour is less likely though, if you can show a "display title" (ex: Hello!) that's different from the "document title" (ex: Home).

This is a common technique people use on their front page—they don't want the title to appear as part of the page content. (See the Hello/Home example above.) I think people tend to do that using `display: none` but we can use a visually-hidden technique to move the text offscreen instead.

**Comment by CreativeDive** [comment]
In reference to my feature request, are there any considerations for the following?

What can we do, if we want to add a custom "Header", which appears above the "Entry Page Title" and the possibility to edit the "Header" like a block inside the block editor?

Which conditions must be fulfilled in the example for a header block?
Always the first position of the block.
--> The block should be positioned above the "Entry Page Title".

Restricting the multiple use of blocks.
--> The block should only be used once without the option to "duplicate" it or add it several times.

Restricting the moving of blocks.
--> The header block should not be movable.

What will the future bring?
Is there a plan about this point in the future?

It would make the editor even better.

This is only an example for a full page header, which is overlaid from the site logo and the site navigation.

![header](https://user-images.githubusercontent.com/31563167/63690797-e0040280-c80d-11e9-9170-dafcf0d01f5c.jpg)

At this time there is no possibility to add custom headers like this in a user friendly way inside the block editor.

**Comment by sarahmonster** [comment]
@CreativeDive It sounds like what you're looking for should be possible once the title is a block (since you could hide it or move it around on the page). 

Logo and site navigation are things that are being worked on in the context of whole-site layouts, but since they're _site_ elements, rather than _page_ elements, you probably wouldn't want them included in a _page_ block like this. Your theme could definitely adjust their positioning so they say on top of a header block though. This "sticky" option is something we've been investigating with respect to the navigation menu block (see #16830).

I wouldn't advocate for adding this much complexity to the core title block itself though, since it would be better provided for by either a custom-developed block via a plugin, or an existing block (the cover block comes to mind).

**Comment by joyously** [comment]
I see this issue has strayed far from the original report, which is about showing the title with the HTML interpreted instead of visible. But showing visible HTML makes it clear to the user that the title actually contains other characters which will affect the user's search and any sorting done by title. It is a big change that should not be taken lightly. 
Themes and core already treat the title differently than other fields, since it can contain HTML that should be interpreted as HTML when output. The user needs to understand that the title is different.

**Comment by karmatosed** [comment]
There has been a little bit of time passed since this issue. I wanted to call attention to the work going on in full site editing that makes all content blocks. With this happening, I think the need for this to be done apart from that might not be the case.

However, it could be great to have some feedback on what examples of HTML are wanted in the title itself.

**Comment by paaljoachim** [comment]
I went ahead and made a new issue in relation to the Block: Title. To give it some more breathing room. https://github.com/WordPress/gutenberg/issues/22621

**Comment by karmatosed** [comment]
@paaljoachim I am not sure if opening a second one when this is still open doesn't distract. Could you perhaps outline why and consider closing this one?

**Comment by joyously** [comment]
Yes, it seems odd to split the conversation.

**Comment by paaljoachim** [comment]
This issue contains a long history of various conversations such as should we move the title to the top bar area. (Full Site Editing has added the title to the top bar.) It is a lot easier to start fresh focusing directly on adding a block title, and keeping this as the history issue. Keeping the fresh issue open and closing this issue. It is also easier getting feedback on a short issue compared to one that is very long.

**Comment by paaljoachim** [comment]
I will go ahead and close this Block: Title issue containing a lot of history.
Lets keep the focus fresh in https://github.com/WordPress/gutenberg/issues/22621 
We can always relate back to specific comments here if/when needed.

**Comment by joyously** [comment]
@paaljoachim the other problem with creating another issue is that the people that commented on the old issue no longer get notified, so you actually reduce the exposure.

**Comment by paaljoachim** [comment]
People who subscribe to this issue will have also seen the various sent lately messages. I just want to make it easier to handle. A short issue is a lot easier to handle then a long. If important information shows up in the new issue I could always add it to this history issue as well.
